### PR TITLE
Add globalization

### PR DESCRIPTION
**Pull request by superkhau**
Connect to strongloop/loopback/issues/2422

### CODE REVIEW

**Comment by superkhau** [comment]
@slnode test please

File: lib/phase-list.js [hunk 1]
```diff
@@ -75,12 +77,12 @@ PhaseList.prototype._resolveNameAndAddToMap = function(phaseOrName) {
...
 
   if (phase.id in this._phaseMap) {
-    throw new Error('Phase "' + phase.id + '" already exists.');
+    throw new Error(g.f('Phase "%s" already exists.', phase.id));
   }
 
 
   if(!phase.__isPhase__) {
-    throw new Error('Cannot add a non phase object to a PhaseList');
+    throw new Error(g.t('Cannot add a non phase object to a PhaseList'));
```
Comment (0candy) [inline lib/phase-list.js]: Add {{}} for {{PhaseList}}

File: package.json [hunk 2]
```diff
@@ -20,7 +20,8 @@
   "license": "MIT",
   "dependencies": {
     "async": "^0.9.0",
-    "debug": "^2.1.0"
+    "debug": "^2.1.0",
+    "strong-globalize": "^2.5.5"
```
Comment (0candy) [inline package.json]: 2.5.6 fixed the swagger-client bug you were seeing yesterday

**Comment by superkhau** [comment]
@0candy Updated, using latest SG also.

File: lib/phase-list.js [hunk 3]
```diff
@@ -75,12 +77,12 @@ PhaseList.prototype._resolveNameAndAddToMap = function(phaseOrName) {
...
 
   if (phase.id in this._phaseMap) {
-    throw new Error('Phase "' + phase.id + '" already exists.');
+    throw new Error(g.f('Phase "%s" already exists.', phase.id));
   }
 
 
   if(!phase.__isPhase__) {
-    throw new Error('Cannot add a non phase object to a PhaseList');
+    throw new Error(g.t('Cannot add a non phase object to a {{PhaseList}}'));
```
Comment (0candy) [inline lib/phase-list.js]: I think you have to run slt-globalize -e again for intl/en/messages.json to get updated again with the brackets change.

**Comment by superkhau** [comment]
@0candy Updated, PTAL again.

**Comment by 0candy** [comment]
The change itself LGTM.
But there is a globalize error right now, all the CI tests fail.
```assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Not supported: en
    at loadGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/lib/globalize.js:354:3)
    at Object.setDefaultLanguage (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/lib/globalize.js:100:3)
    at new StrongGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/index.js:27:41)
    at StrongGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/index.js:25:12)
    at Object.<anonymous> (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/lib/providers/apns.js:11:9)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
```
So please don't merge yet.

**Comment by 0candy** [comment]
Please fix this too https://github.com/strongloop/loopback-component-push/pull/118#issuecomment-235120080

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
@0candy the strong-globalize problem has been resolved. This PR is now failing on one of the tests, which looks like it needs to be updated to accept the translated string:
```
  29 passing (54ms)
  1 failing

  1) PhaseList phaseList.addAfter throws when the "after" phase was not found:
     AssertionError: expected [Function] to throw error matching /unknown-phase/ but got 'Unknown phase: %'
      at Context.<anonymous> (test/phase-list.test.js:146:18)
```

**Comment by 0candy** [comment]
@superkhau Please fix the line that @rmg pointed out. Thanks.

**Comment by superkhau** [comment]
@gunjpan @richardpringle PTAL, updated with latest conventions.

**Comment by superkhau** [comment]
@rmg Tons of unrelated failures in this PR. Can you provide some insight? It seems like some NPM registry doesn't contain `loopback` according to the errors.

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
@superkhau they look like they are probably related to high load. Looking at the file dump from the ephemeral registry they were using, there were 6 partial copies of loopback mid-cache, so I think it just sort of choked on all the downstream jobs running at the same time.

On the upside, if you look at the `cis-jenkins` downstream jobs, those are the ones that are actually version aware, so only downstream modules that actually use this version of loopback-phase were triggered. And those all passed :-)

**Comment by superkhau** [comment]
>On the upside, if you look at the cis-jenkins downstream jobs, those are the ones that are actually version aware, so only downstream modules that actually use this version of loopback-phase were triggered. And those all passed :-)

:+1: That's a definition plus as the ones actually failing are unrelated and the ones I care about are passing on `cis-jenkins`. Thanks for checking, I'm gonna merge as I feel this is safe enough. ;)

File: lib/phase-list.js [hunk 4]
```diff
@@ -110,7 +112,7 @@ PhaseList.prototype.addAt = function(index, phase) {
 PhaseList.prototype.addAfter = function(after, phase) {
   var ix = this.getPhaseNames().indexOf(after);
   if (ix === -1) {
-    throw new Error('Unknown phase: "' + after + '"');
+    throw new Error(g.f('Unknown phase: %', after));
```
Comment (gunjpan) [inline lib/phase-list.js]: @superkhau : this should be:
`throw new Error(g.f('Unknown phase: %s', after));`

File: test/phase-list.test.js [hunk 5]
```diff
@@ -143,7 +143,7 @@ describe('PhaseList', function() {
 
     it('throws when the "after" phase was not found', function() {
       expect(function() { phaseList.addAfter('unknown-phase', 'a-name'); })
-        .to.throw(/unknown-phase/);
+        .to.throw('Unknown phase: %');
```
Comment (gunjpan) [inline test/phase-list.test.js]: use a regex to match important parts from the string:
```javascript
  ...
    .to.throw(/unknown phase/i);
```

**Comment by gunjpan** [comment]
@superkhau: Reviewed. Few nit-picks, otherwise, LGTM.

**Comment by superkhau** [comment]
@gunjpan @richardpringle Updated. PTAL again.

**Comment by richardpringle** [comment]
LGTM

**Pull request merged**